File: tests/r_parity/83_lpdid.py
"""StatsPAI LP-DiD parity (Python side) -- Module 83.

Runs ``sp.lp_did`` on a staggered fixture. The companion ``83_lpdid.R``
transcribes the same estimator directly rather than calling a package,
because no LP-DiD package is installed on the R side; the transcription
follows the definition ``sp.lp_did`` documents and implements:

  * at horizon h, regress  Δy = Y_{t+h} − Y_{t−1}  on  Δd = d_t − d_{t−1};
  * the treated arm is Δd == 1;
  * a clean control must have d == 0 across the whole window
    [t + min(−1, h−1), t + max(0, h)] and Δd == 0;
  * calendar-time fixed effects, cluster-robust SE by unit.

The fixture deliberately contains **only 0→1 transitions**. ``sp.lp_did``
documents that its handling of switch-off events is "not verified against
the published paper", so pinning a reference against that branch would
pin behaviour the maintainer has explicitly flagged as unconfirmed. That
branch stays an open coverage item rather than a silently blessed one.

Tolerance: rel < 1e-10. Both sides solve the same OLS on the same rows,
so only floating-point summation order should separate them.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "83_lpdid"
HORIZONS = (-2, 3)
# h = -1 is the LP-DiD base period and is identically zero by construction,
# so it carries no information and is excluded from the comparison.
COMPARED_HORIZONS = (-2, 0, 1, 2, 3)


def make_data(n_units: int = 200, n_periods: int = 10, seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    cohorts = rng.choice([4, 6, 8, 0], size=n_units, p=[0.25, 0.25, 0.25, 0.25])
    rows = []
    for i in range(n_units):
        g = int(cohorts[i])
        unit_fe = rng.normal(0.0, 1.0)
        for t in range(1, n_periods + 1):
            treated = 1 if (g > 0 and t >= g) else 0
            effect = 0.8 * (t - g + 1) if treated else 0.0
            rows.append(
                {
                    "unit": i,
                    "time": t,
                    "d": treated,
                    "y": unit_fe + 0.3 * t + effect + rng.normal(0.0, 0.5),
                }
            )
    return pd.DataFrame(rows)


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    fit = sp.lp_did(df, y="y", unit="unit", time="time", treatment="d", horizons=HORIZONS)
    es = pd.DataFrame(fit.model_info["event_study"])

    rows: list[ParityRecord] = []
    for h in COMPARED_HORIZONS:
        row = es.loc[es["relative_time"] == h]
        if row.empty:
            raise SystemExit(f"lp_did returned no horizon {h}")
        row = row.iloc[0]
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"lpdid_h{h}_att",
                estimate=float(row["att"]),
                se=float(row["se"]),
                n=int(row["n_obs"]),
            )
        )

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "clean_controls": "not_yet_treated",
            "time_fe": True,
            "cluster": "unit",
            "horizons": list(HORIZONS),
            "reference_kind": "direct_transcription_no_r_package",
            "note": (
                "Fixture has only 0->1 transitions; sp.lp_did documents its "
                "switch-off handling as not verified against the published paper, "
                "so that branch is deliberately not pinned here."
            ),
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/83_lpdid.R
# StatsPAI LP-DiD parity (R side) -- Module 83.
#
# No LP-DiD package is installed on this R side, so this is a DIRECT
# TRANSCRIPTION of the estimator sp.lp_did documents and implements,
# written independently from the Python code:
#
#   at horizon h, regress  dy = Y_{t+h} - Y_{t-1}  on  dd = d_t - d_{t-1}
#     * treated arm      : dd == 1
#     * clean control    : d == 0 across [t + min(-1, h), t + max(0, h)] (Stata lpdid: CCS_m<h> = CCS_0)
#                          and dd == 0
#     * calendar-time fixed effects
#     * cluster-robust SE by unit
#
# Reading the same CSV bytes as the Python side, the two must agree to
# floating-point noise: they are the same OLS on the same rows.
#
# Tolerance: rel < 1e-10.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(sandwich)
  library(lmtest)
})

MODULE <- "83_lpdid"
COMPARED_HORIZONS <- c(-2, 0, 1, 2, 3)

df <- read_csv_strict(MODULE)
df <- df[order(df$unit, df$time), ]

units <- unique(df$unit)
n_periods <- length(unique(df$time))

# Per-unit treatment path, indexed by period position.
paths <- split(df$d, df$unit)
ys <- split(df$y, df$unit)

lpdid_at_horizon <- function(h) {
  win_lo <- min(-1L, h)
  win_hi <- max(0L, h)

  recs <- list()
  for (u in units) {
    d_u <- paths[[as.character(u)]]
    y_u <- ys[[as.character(u)]]
    n <- length(d_u)
    for (k in seq_len(n)) {
      # k is the position of period t (1-based). Y_{t-1} is k-1, Y_{t+h} is k+h.
      base_idx <- k - 1L
      fut_idx <- k + h
      if (base_idx < 1L || fut_idx < 1L || fut_idx > n) next
      dy <- y_u[fut_idx] - y_u[base_idx]
      dd <- d_u[k] - d_u[base_idx]

      lo <- k + win_lo
      hi <- k + win_hi
      stable_zero <- FALSE
      if (lo >= 1L && hi <= n) {
        stable_zero <- all(d_u[lo:hi] == 0)
      }

      keep <- (dd == 1) || (stable_zero && dd == 0)
      if (!keep) next
      recs[[length(recs) + 1L]] <- data.frame(
        unit = u, time = k, dy = dy, dd = max(dd, 0), stringsAsFactors = FALSE
      )
    }
  }
  if (length(recs) == 0L) return(NULL)
  do.call(rbind, recs)
}

rows <- list()
for (h in COMPARED_HORIZONS) {
  s <- lpdid_at_horizon(h)
  if (is.null(s)) stop(sprintf("empty LP-DiD sample at horizon %d", h))

  fit <- lm(dy ~ dd + factor(time), data = s)
  # sp.lp_did applies the Stata-style cluster correction
  #     G/(G-1) * (N-1)/(N-K)
  # vcovCL(cadjust = TRUE) supplies only the first factor, so the second is
  # applied here explicitly. Without it the two sides differ by ~0.4% on this
  # fixture -- small enough to look like noise and wrong for a stated reason.
  vc <- sandwich::vcovCL(fit, cluster = s$unit, type = "HC0", cadjust = TRUE)
  n_obs <- nrow(s)
  k_par <- fit$rank
  vc <- vc * ((n_obs - 1) / (n_obs - k_par))
  ct <- lmtest::coeftest(fit, vcov. = vc)

  rows[[length(rows) + 1L]] <- parity_row(
    module    = MODULE,
    statistic = sprintf("lpdid_h%d_att", h),
    estimate  = unname(ct["dd", "Estimate"]),
    se        = unname(ct["dd", "Std. Error"]),
    n         = nrow(s)
  )
}

write_results(MODULE, rows,
              extra = list(reference_kind = "direct_transcription_no_r_package",
                           clean_controls = "not_yet_treated",
                           time_fe = TRUE,
                           cluster = "unit"))
